### PR TITLE
feat(ECC): add ECC for coupledL2

### PR DESCRIPTION
**Pull request by Ma-YX**
add optional ECC for tag & data

1. ECC SRAM
1.1 merge tag/data SRAM with corresponding ECC SRAM
1.2 For data ECC, currently one cache line(512 bits) is separate into 4 parts to encode/decode, under the request of data-split of data SRAM, which might be changed later 

2. ECC error
2.1 ECC error in L2 will report to beu; and will set corrupt/deny or RespErr correspondingly to L1/L3 (TL or CHI)
2.2 error from L1/L3 will be kept in L2 directory(meta), no error reporting

3. Data Check & Poison
3.1 Data Check is only apply to data(optional) currently, using odd parity check or SECDED (default odd parity)
3.2 Poison is simply applied: only data poison & over poison based on ECC

// Hint: l2-beu error report datapath might be changed as optional

### CODE REVIEW

**Review by linjuanZ — CHANGES_REQUESTED** [review]

File: src/main/scala/coupledL2/CoupledL2.scala [hunk 1]
```diff
@@ -452,6 +467,36 @@ abstract class CoupledL2Base(implicit p: Parameters) extends LazyModule with Has
...
+        case (s, i) => s.io.error
+      }
+      l2ECCArb.io.in <> VecInit(slices_l2ECC)
+      l2ECCArb.io.out.ready := true.B
+      io.error.valid := l2ECCArb.io.out.fire && l2ECCArb.io.out.bits.valid
+      io.error.address := l2ECCArb.io.out.bits.address
+    } else {
+      io.error.valid := false.B
+      io.error.address := 0.U.asTypeOf(io.error.address)
+    }
```
[the remark this answers is not in the document]
Comment (Ma-YX) [reply src/main/scala/coupledL2/CoupledL2.scala:498]: This is caused by not fully deleting duplicated code during two rebase merge, I'll delete one later

**Review by linjuanZ — COMMENTED** [review]
Please make sure OpenLLC handles DataCheck correctly before this pr is merged. @Kumonda221-CrO3

File: src/main/scala/coupledL2/CoupledL2.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (Ma-YX) [reply src/main/scala/coupledL2/CoupledL2.scala]: fixed in 495d96f31b8a1a433fd4115e1bff3df4ee32db2b

File: src/main/scala/coupledL2/DataStorage.scala [hunk 2]
```diff
@@ -70,14 +95,44 @@ class DataStorage(implicit p: Parameters) extends L2Module {
   val wen = io.req.valid && io.req.bits.wen
   val ren = io.req.valid && !io.req.bits.wen
 
+  val arrayWrite = Wire(new DSECCBankBlock)
+  val arrayWriteData = if (enableDataECC) {
+    // cacheParams.dataCode.encode(io.wdata.data).pad(encDataPaddingBits)
+    Cat(VecInit(Seq.tabulate(4)(i => io.wdata.data((blockBytes * 2) * (i + 1) - 1, (blockBytes * 2) * i))).map(data => cacheParams.dataCode.encode(data)))
```
[the remark this answers is not in the document]
Comment (Ma-YX) [reply src/main/scala/coupledL2/DataStorage.scala:101]: fixed in 9e575788365bc8981a528bc7f71781f73676d6ac

File: src/main/scala/coupledL2/DataStorage.scala [hunk 3]
```diff
@@ -70,14 +95,44 @@ class DataStorage(implicit p: Parameters) extends L2Module {
...
   // TODO: This check is done elsewhere
-  array.io.w.apply(wen, io.wdata, arrayIdx, 1.U)
+  array.io.w.apply(wen, arrayWrite, arrayIdx, 1.U)
   array.io.r.apply(ren, arrayIdx)
 
+
+  //  val eccData = arrayRead.data(encDataBits - 1, 0)
+  val error = if (enableDataECC) {
+    // cacheParams.dataCode.decode(eccData).error && RegNext(RegNext(io.req.valid && !io.req.bits.wen))
+    VecInit(Seq.tabulate(4)(i => arrayRead.data(encDataBankBits * (i + 1) - 1, encDataBankBits * i))).
```
[the remark this answers is not in the document]
Comment (Ma-YX) [reply src/main/scala/coupledL2/DataStorage.scala]: fixed in 9e575788365bc8981a528bc7f71781f73676d6ac

File: src/main/scala/coupledL2/CoupledL2.scala [hunk 4]
```diff
@@ -452,6 +467,36 @@ abstract class CoupledL2Base(implicit p: Parameters) extends LazyModule with Has
...
+    // ECC error
+    if (enableECC) {
+      val l2ECCArb = Module(new Arbiter(new L2CacheErrorInfo(), slices.size))
+      val slices_l2ECC = slices.zipWithIndex.map {
+        case (s, i) => s.io.error
+      }
+      l2ECCArb.io.in <> VecInit(slices_l2ECC)
+      l2ECCArb.io.out.ready := true.B
+      io.error.valid := l2ECCArb.io.out.fire && l2ECCArb.io.out.bits.valid
+      io.error.address := l2ECCArb.io.out.bits.address
```
[the remark this answers is not in the document]
Comment (Ma-YX) [reply src/main/scala/coupledL2/CoupledL2.scala]: fixed in 27566879e80f05f21079c63108b9d1eef28ce501 and 57b4a63dcb50047225754593b70b11d3e5e71a71

File: src/main/scala/coupledL2/tl2chi/TXDAT.scala [hunk 5]
```diff
@@ -98,6 +98,7 @@ class TXDAT(implicit p: Parameters) extends TL2CHIL2Module {
 
   io.out.valid := taskValid
   io.out.bits := toCHIDATBundle(taskR.task, beat, beatsOH)
+  io.out.bits.respErr := Mux(taskR.task.corrupt, Mux(taskR.task.denied, RespErrEncodings.NDERR, RespErrEncodings.DERR), RespErrEncodings.OK)
```
[the remark this answers is not in the document]
Comment (Ma-YX) [reply src/main/scala/coupledL2/tl2chi/TXDAT.scala]: fixed in 57b4a63dcb50047225754593b70b11d3e5e71a71

File: src/main/scala/coupledL2/tl2chi/chi/Message.scala [hunk 6]
```diff
@@ -299,7 +299,8 @@ trait HasCHIMsgParameters {
   def DATAID_WIDTH = 2
   def BE_WIDTH = DATA_WIDTH / 8
   def DATA_WIDTH = 256
-  def DATACHECK_WIDTH = DATA_WIDTH / 8
+  def DATACHECK_WIDTH = DATA_WIDTH / 64
```
[the remark this answers is not in the document]
Comment (Ma-YX) [reply src/main/scala/coupledL2/tl2chi/chi/Message.scala]: fixed in 57b4a63dcb50047225754593b70b11d3e5e71a71

File: src/main/scala/coupledL2/Common.scala [hunk 7]
```diff
@@ -428,3 +429,8 @@ class PCrdGrantMatcher(val numPorts: Int) extends Module {
     p.pCrdType === io.rxrsp.bits.pCrdType
   }).asUInt
 }
+
+class L2CacheErrorInfo(implicit p: Parameters) extends L2Bundle {
+  val valid = Bool()
+  val address = UInt(48.W)
```
[the remark this answers is not in the document]
Comment (Ma-YX) [reply src/main/scala/coupledL2/Common.scala]: fixed in 8514d3a0a99b66c26b97d54c32848b302c889bf6

**Review by linjuanZ — APPROVED** [review]

**Comment by Kumonda221-CrO3** [comment]
Test passed on OpenLLC with ```dataCheck = Some("oddparity")```.

**Review by Kumonda221-CrO3 — APPROVED** [review]

**Pull request merged**